udp.go

```diff
@@ -9,6 +9,34 @@ import (
 	"golang.org/x/net/ipv6"
 )
 
+type SessionUDPFactory interface {
+	// InitConn set's up 'conn' to be used with a SessionUDP.
+	// Must be called before 'conn' is passed to SessionUDP.ReadRequest()
+	InitConn(conn *net.UDPConn) error
+
+	// CreateSessionUDP creates a SessionUDP object which can manage the state
+	// of a single UDP transaction at a time.
+	// Multiple SessionUDP objects may be created on the same underlying 'conn' by
+	// calling this multiple times with the same 'conn'.
+	CreateSessionUDP(msgSize int) SessionUDP
+}
+
+// SessionUDP holds manages a UDP Request/Response transaction.
+type SessionUDP interface {
```

Comment:
I'm not convinced this is needed either. You can use the `Writer` interface to provide an entirely custom implementation that doesn't call into the default `WriteToSessionUDP`. Also `RemoteAddr` and `LocalAddr` are only used by `ResponseWriter` which is an interface and can easily be wrapped by middleware to provide your own custom implementations of these as well.

Reply:
Again, I don't see how a `ResponseWriter` wrapper would get access to `SessionUDP`, which stored to the private (default) implementation `response`?

Could you have a look at my implementation of the SessionUDP interface proposed here (at https://github.com/cilium/cilium/blob/pr/redirect-with-tproxy/pkg/fqdn/dnsproxy/udp.go) and give some more concrete hints how to do this like you propose above?